tools/cxbe/Xbe.cpp

```diff
@@ -513,8 +514,12 @@ Xbe::Xbe(class Exe *x_Exe, const char *x_szTitle, bool x_bRetail, const std::vec
                 printf("Xbe::Xbe: Generating Section %.04X...", v);
 
                 uint32 RawSize = m_SectionHeader[v].dwSizeofRaw;
+                uint32 VirtSize = m_SectionHeader[v].dwVirtualSize;
+                uint32 maxSize = std::max(VirtSize, RawSize);
 
-                m_bzSection[v] = new uint08[RawSize];
+                m_bzSection[v] = new uint08[maxSize];
+
+                memset(m_bzSection[v], 0, maxSize);
```

Comment:
imo, this should be after copy raw section (couple lines below) then perform `memset(m_bzSection[v] + RawSize, 0, VirtSize);`. This way, it will reduce unnecessary set the memory twice. Raw section is always written to memory. Whilst virtual block of the section is always set to 0.

Reply:
That would cause an out of bounds write. The virtual size is not a block of zeroes, it's the size of the section when it's loaded into memory.

Reply:
I forgot to include ` - RawSize` after VirtSize, so it has been edited. Anyway, yes the section in virtual memory is VirtSize, but the RawSize is from data stored inside xbe file. When xbe's section is load into memory VirtSize is allocated, only first part of the raw section from xbe is written to memory. The remaining size, aka `VirtSize - RawSize` are set to zero'd.

Reply:
We can't do that either, there's no guarantee that the raw size is smaller than the virtual size. In fact the exe produced by lld contains sections where this is not the case.

Reply:
Then wouldn't that be invalid exe binary? Otherwise it wouldn't work on any OSes, including Windows/Xbox. Raw size is always equal or less than virtual size no matter what.

Reply:
No, having the raw size larger than the virtual size is perfectly fine and does work.

Reply:
> The size of the section (for object files) or the size of the initialized data on disk (for image files). For executable images, this must be a multiple of FileAlignment from the optional header. If this is less than VirtualSize, the remainder of the section is zero-filled. Because the SizeOfRawData field is rounded but the VirtualSize field is not, it is possible for SizeOfRawData to be greater than VirtualSize as well. When a section contains only uninitialized data, this field should be zero.

source: https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#section-table-section-headers

Ah, I see raw size could be a bit larger than virtual size. Never mind then. (You can close this discussion as resolve.)